rfcs/0005-nix-encryption.md

```diff
@@ -0,0 +1,273 @@
+  expected by `encryptString` and `nix decrypt`.
+
+A typical usage in a module NixOS would look like this:
+
+```
+let
+  configFile = pkgs.writeText "foo.conf"
+    ''
+      # Store the password of the foo service in encrypted form.
+      password=${builtins.encryptString <nixos-store-key> cfg.password}
+    '';
+in {
+
+  systemd.services.foo = {
+    preStart =
+      ''
+        # Decrypt the configuration file.
+        ${config.nix.package}/bin/nix-store \
+          --decrypt ${toString <nixos-store-key>} \
+          ${configFile} > /run/secret/foo.conf
+      '';
+    # Run the service using the decrypted configuration file/
+    serviceConfig.Exec =
+      "${pkgs.foo}/bin/foo -c /run/secret/foo.conf";
+  };
+};
+```
+(See below for a discussion of simplifications to this example.)
+
+At evaluation time, this causes the following string to be emitted by
+the Nix evaluator:
+```
+password=<{|nixcrypt:base64-encoded-data|}>
+```
+
+Note that we don't necessarily encrypt entire files (though that's
+possible); generally, we'll encrypt only the sensitive parts of
+files. This allows users to read non-sensitive parts of configuration
+files. It *may* also enable graceful degradation if secrets cannot be
+decrypted, but this is risky (see below under "Decryption failure").
+
+The assumption is that `<nixos-store-key>` resolves to something like
+`/var/lib/nixos/store-key`, which should contain a key generated by
+`nix generate-key` and should obviously be readable by root only.
+
+# Prototype
+
+* Nix implementation: https://github.com/edolstra/nix/commit/6b70036
+* Nixpkgs example: https://github.com/edolstra/nixpkgs/commit/4c8212069429bf9fb959e00ce8d9345ac7cb7ff0
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+* The main downside of this approach is that encrypted files need to
+  be decrypted before they can be used. Thus, for example, a daemon
+  configuration file cannot be used directly from the Nix store by
+  passing something like `--config /nix/store/.../foo.conf` to the
+  daemon. Instead, we have to decrypt the configuration file to some
+  suitably secure temporary location in a pre-start script, and then
+  pass that location (e.g. `--config /tmp/.../foo.conf`).
+
+  In NixOS, we can provide some helper functions to reduce boilerplate
+  code dealing with encrypted files in systemd services. For example,
+  we can add an option to allow modules to declare files that need
+  decryption:
+
+    ```
+    security.encryptedFiles = [ wpaSupplicantConf ];
+    ```
+
+  which would be decrypted by the activation script to a well-known
+  location such as `/var/secrets/<storepath>`. This way, services can
+  refer to the decrypted path easily:
+
+    ```
+    systemd.services.wpa_supplicant.serviceConfig.Exec =
+      "wpa_supplicant -c /var/secrets/${wpaSupplicantConf}";
+      # i.e. ... -c /var/secrets/nix/store/.../wpa_supplicant.conf
+    ```
+
+  The above can also be extended to support access by non-root users,
+  e.g.
```

Comment:
This seems to be a critically important component, so individual services don't need (poorly) re-implement this behavior.